File: src/tools.js
/**
 * tools
 */

const http = require('http'); 
var fs = require('fs');
const child_process = require('child_process');
const path = require('path');
const readline = require('readline');

//////////////////////////////////////////////////////

// First, checks if it isn't implemented yet.
if (!String.prototype.format) {
    String.prototype.format = function() {
      var args = arguments;
      return this.replace(/{(\d+)}/g, function(match, number) { 
        return typeof args[number] != 'undefined'
          ? args[number]
          : match
        ;
      });
    };
  }

//////////////////////////////////////////////////////

let js = {}

js.is_boolean = (value) => {
    return typeof value == 'boolean';
}

js.is_number = (value) => {
    return typeof value == 'number';
}

js.is_string = (value) => {
    return typeof value == 'string';
}

js.is_text = (val) => {
    return typeof val == 'string' && val.trim().length > 0;
}

js.is_array = (value) => {
    return Array.isArray(value);
}

js.is_object = (value) => {
    return !Array.isArray(value) && typeof value == 'object';
}

js.print = (...args) => {
    console.log(...args);
}

js.trace = (...args) => {
    console.log(args.join(' '));
    console.log(new Error().trace);
}

js.assert = (val, ...args) => {
    if (!val) throw new Error(args.join(' '));
}
js.error = (...args) => {
    throw new Error(args.join(' '));
}

//////////////////////////////////////////////////////

const files = {};

files.delete = (path) => {
    try { fs.unlinkSync(path); } catch (e) { }
}

files.mk_folder = (path) => {
    if (!fs.existsSync(path)){
        fs.mkdirSync(path, {recursive: true});
    }
}

files.is_folder = (path) => {
    try {
        return fs.lstatSync(path).isDirectory();
    } catch (e) {
        return false;
    }
}

files.is_file = (path) => {
    try {
        return fs.lstatSync(path).isFile(); 
    } catch (e) {
        return false;
    }
}

files.is_exist =(path) => {
    return fs.existsSync(path);
}

files.write = (path, content, encoding) => {
    fs.writeFileSync(path, content, {encoding: encoding});
}

files.read = (path, encoding = 'utf-8') => {
    return fs.readFileSync(path, {encoding: encoding});
}

files.copy = (_from, _to) => {
    fs.copyFileSync(_from, _to);
}

files.size = (_path) => {
    try {
        var stats = fs.statSync(_path);
        var bytes = stats.size;
        return bytes;
    } catch (e) {
        return -1;
    }
}

//////////////////////////////////////////////////////

const terminal = {}

async function read_line(msg) {
    let promise = new Promise((resolve, reject) => {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        rl.question(msg || "input:", (answer) => {
            rl.close();
            resolve(answer);
        });
    })
    return promise;
}
terminal.read_line = read_line;

async function read_selection(msg, options) {
    console.log(msg || "select:", options);
    while(true) {
        let txt = await read_line();
        if (condition) {
            if (Array.find(options, txt)) {
                return txt
            }
        }
    }
}
terminal.read_selection = read_selection;

//////////////////////////////////////////////////////

const tools = {};

tools.is_windows = () => {
    return process.platform == "win32" || process.platform == "win64";
}

tools.is_mac = () => {
    return process.platform == "darwin";
}

tools.is_linux = () => {
    return process.platform == "linux";
}

tools.execute = (cmd) => {
    try {
        let options = {stdio : 'pipe' };
        const output = child_process.execSync(cmd, options);
        return [true, output.toString()];
    } catch (error) {
        return [false, `${error}`];
    }
}

tools.spawn = (cmd, args = [], options = {}) => {
    if (!options) {
        options = {};
    }
    let _options = {
        cwd: process.cwd(),
        env: process.env,
        stdio: [process.stdin, process.stdout, process.stderr],
        encoding: 'utf-8'
    }
    Object.assign(_options, options)
    var child = child_process.spawnSync(cmd, args, _options);
    return [child.error == null, child.error, child];
}

tools.get_separator = () => {
    return tools.is_windows() ? "\\" : "/";
}

tools.parse_path = (_path) => {
    let parsed = path.parse(_path);
    let dir = parsed.dir;
    let name = parsed.name;
    let ext = parsed.ext.split(".").pop();
    let nameWithExt = parsed.base;
    return [dir, name, ext, nameWithExt];
}

//////////////////////////////////////////////////////

let encryption = {}

encryption.base64_encode = (content) => {
    return Buffer.from(content).toString("base64");
}

encryption.base64_decode = (content) => {
    return Buffer.from(content, "base64").toString('ascii');
}

//////////////////////////////////////////////////////

let httpy = {}

function httpy_curl(url, method, params, headers) {
    //
    js.assert(js.is_text(url));
    method = method.toUpperCase();
    js.assert(method == 'POST' || method == 'GET');
    params = params || {};
    headers = headers || {};
    //
    const tempFile = './.js.http.log';
    files.delete(tempFile);
    //
    let h = '';
    for (const [k, v] of Object.entries(headers)) {
        assert(is_text(k));
        assert(is_string(v) || is_number(v));
        if (h !== '') {
            h += ';';
        }
        h += `-H '${String(k)}:${String(v)}'`;
    }
    //
    let b = '';
    if (method === 'GET') {
    for (const [k, v] of Object.entries(params)) {
        if (!url.includes('?')) {
            url += '?';
        }
        assert(is_text(k));
        assert(is_string(v) || is_number(v));
        url += `${String(k)}=${String(v)}`;
    }
    } else if (method === 'POST') {
        b = `-d '${JSON.stringify(params)}'`;
    }
    //
    const cmd = `curl "${url}" -i --silent  -X ${method} -o "${tempFile}" ${h} ${b}`;
    const [isOk, output] = tools.execute(cmd);
    const content = files.read(tempFile, 'utf8');
    files.delete(tempFile);
    if (!isOk) {
        return [-1, output];
    }
    //
    const [head, body] = content.split(/\r?\n\r?\n/g, 2);
    const codeMatch = head.match(/HTTP.+\s(\d{3})/);
    const code = codeMatch ? parseInt(codeMatch[1]) : -1;
    //
    if (code < 0) {
        return [-1, output];
    } else {
        return [code, body];
    }
}

httpy.request = (url, method, params, headers) => {
    const [code, content] = httpy_curl(url, method, params, headers);
    return [code == 200, code, content];
}
  
httpy.get = (url, params, headers) => {
    return httpy.request(url, 'GET', params, headers);
}
  
httpy.post = (url, params, headers) => {
    return httpy.request(url, 'POST', params, headers);
}

httpy.download = (url, _path, withPrint = false) => {
    if (!js.is_text(url) || !js.is_text(_path)) {
      throw new Error('Invalid URL or file path');
    }
    const folder = path.dirname(_path);
    fs.mkdirSync(folder, { recursive: true });
    let cmd = `curl -L "${url}" -o "${_path}" --max-redirs 3`;
    try {
        let options = {stdio : withPrint ? null : 'pipe' };
        const output = child_process.execSync(cmd, options);
        return [true, output.toString()];
    } catch (error) {
        return [false, `${error}`];
    }
}

//////////////////////////////////////////////////////

module.exports = {js, files, terminal, tools, encryption, httpy};

File: src/tools.py
"""
tools
"""

import http
import os
import subprocess
import pathlib
import base64
import json
import platform
import urllib
import urllib.request
import traceback
import shutil

######################################################

class Globals:
    pass

Globals.createFunc = None
Globals.originsPath = os.path.join(os.path.dirname(__file__), "../resources/origins.yaml")

class KEYS:
    NAME = "NAME"
    TYPE = "TYPE"
    URL = "URL"
    EXT = "EXT"
    BRANCH = "BRANCH"
    DIR_I = "DIR_I"  # -I
    DIR_L = "DIR_L"  # -L
    LIB_L = "LIB_L"  # -l
    FLAGS = "FLAGS"  # flags
    FILES = "FILES"  # include source files in lib
    WIN = "WIN"
    LNX = "LNX"
    MAC = "MAC"
    pass

class TYPES:
    GIT = "GIT"
    ZIP = "ZIP"
    GZ = "GZ"
    pass

######################################################

def py_implode(args, separator=" "):
    return separator.join(map(str, args))

def py_explode(text, separator=" "):
    return text.split(separator, -1)

class py:
    is_boolean = lambda value: isinstance(value, bool)
    is_number = lambda value: isinstance(value, (int, float))
    is_string = lambda value: isinstance(value, str)
    is_text = lambda val: isinstance(val, str) and val.strip()
    is_array = lambda value: isinstance(value, list)
    is_object = lambda value: isinstance(value, dict)
    is_function = lambda val: callable(val)
    implode = lambda separator, *args: py_implode(args, separator)
    explode = lambda separator, text: py_explode(text, separator)
    print = lambda *args: print(*args)
    trace = lambda *args: (print(py_implode(args)), traceback.print_stack())
    check = lambda val, *args: (_ for _ in ()).throw(AssertionError(py_implode(args))) if not val else None
    error = lambda *args: (_ for _ in ()).throw(Exception(py_implode(args)))
    pass

######################################################

def files_write(path, content, encoding=None):
    return open(path, 'w' if encoding != None else "wb", encoding=encoding).write(content)


def files_read(path, encoding=None):
    return open(path, 'r' if encoding != None else "rb", encoding=encoding).read()

class files:
    delete = lambda path: os.remove(path) if os.path.exists(path) else None
    mk_folder = lambda path: os.makedirs(path, exist_ok=True)
    is_folder = lambda path: os.path.isdir(path)
    is_file = lambda path: os.path.isfile(path)
    is_exist = lambda path: os.path.exists(path)
    write = files_write 
    read = files_read
    copy = lambda _from, _to: shutil.copyfile(_from, _to)
    size = lambda _path: os.path.getsize(_path) if os.path.exists(_path) else -1

######################################################

async def read_line(msg="input:"):
    return input(msg)

async def read_selection(options, msg="select:"):
    print(msg, options)
    while True:
        txt = await read_line()
        if txt in options:
            return txt
        else:
            print("invalid!")

class terminal:
    read_line = read_line
    read_selection = read_selection
    pass

######################################################

def tools_execute(cmd, args=[], cwd=None, encoding='gbk'):
    try:
        p = subprocess.check_output(
            cmd,
            shell = True,
            cwd = cwd,
            encoding = encoding,
            stderr = subprocess.STDOUT
        )
        return [True, p]
    except subprocess.CalledProcessError as e:
        return [False, "cmd:<{}> code:({}) msg:{}".format(e.cmd, e.returncode, e.output)]
    except Exception as e:
        return [False, e]
    
def tools_spawn(cmd, args=[], cwd=None, encoding='gbk'):
    try:
        p = subprocess.Popen(
            [cmd] + args,
            encoding = encoding,
            cwd = cwd
        )
        p.wait()
        return [True, p.returncode]
    except subprocess.CalledProcessError as e:
        return [False, "cmd:<{}> code:{} msg:{}".format(e.cmd, e.returncode, e.output)]
    except Exception as e:
        return [False, e]
        

def tools_parse_path(_path):
    _path = pathlib.Path(_path)
    return (_path.parent, _path.stem, _path.suffix.lstrip('.'), _path.name)

class tools:
    is_windows = lambda: platform.system() == 'Windows'
    is_mac = lambda: platform.system() == 'Darwin'
    is_linux = lambda: platform.system() == 'Linux'
    execute = tools_execute
    spawn = tools_spawn
    get_separator = lambda: "/" if tools.is_windows() else "\\"
    parse_path = tools_parse_path
    validate_path = lambda path: path.replace("/", tools.get_separator()).replace("\\", tools.get_separator())
    pass

######################################################

class encryption:
    base64_encode = lambda content: base64.b64encode(content.encode()).decode()
    base64_decode = lambda content: base64.b64decode(content).decode()
    pass

######################################################

def httpy_request(url, isPost, params={}, headers={}):
    _method = "POST" if isPost else "GET"
    _params = urllib.parse.urlencode(params)
    _url = url if isPost else f"{url}?{_params}"
    _data = _params.encode() if isPost else None
    _request = urllib.request.Request(_url, data=_data, method=_method)
    try:
        _response = urllib.request.urlopen(_request)
        code = _response.status
        isOk = code >= 200 and code < 300
        data = _response.read()
        return [isOk, code, data]
    except urllib.error.HTTPError as e:
        return [False, e.code, "url:<{}> msg:{}".format(e.filename, e.msg)]
    except Exception as e:
        return [False, -1, e]

def httpy_download(url, _path):
    try:
        _response = urllib.request.urlopen(url)
        _code = _response.getcode()
        if _code != 200:
            return [False, _code, _response.msg]
        file_contents = _response.read()
        file = open(_path, "wb")
        file.write(file_contents)
        return [True, _code, _response.msg]
    except urllib.error.URLError as e:
        return [False, e.code, "url:<{}> msg:{}".format(e.filename, e.msg)]
    except PermissionError as e:
        return [False, e.errno, "url:<{}> msg:{}".format(e.filename, e.strerror)]
    except Exception as e:
        return [False, -1, e]

class httpy:
    get = lambda url, params={}, headers={}: httpy_request(url, False, params, headers)
    post = lambda url, params={}, headers={}: httpy_request(url, True, params, headers)
    download = httpy_download
    pass

######################################################
